test_conformance/generic_address_space/atomic_tests.cpp

```diff
@@ -0,0 +1,231 @@
+// Every other workgroup will use either a global or local
+// pointer on an atomic operation.
+const char* KernelSourceInvariant = R"OpenCLC(
+kernel void testKernel(global int* globalPtr, local int* localPtr) {
+    int gid = get_global_id(0);
+    int tid = get_local_id(0);
+    int wgid = get_group_id(0);
+    int wgsize = get_local_size(0);
+
+    if (tid == 0) *localPtr = 0;
+
+    barrier(CLK_LOCAL_MEM_FENCE);
+
+    // Initialise the generic pointer to
+    // the global.
+    generic int* ptr = globalPtr + wgid;
+
+    // In a workgroup-invariant way, select a localPtr instead.
+    if ((wgid % 2) == 0)
+        ptr = localPtr;
+
+    int inc = atomic_fetch_add(ptr, 1);
+
+    // In the cases where the local memory ptr was used,
+    // save off the final value.
+    if ((wgid % 2) == 0 && inc == (wgsize-1))
+        globalPtr[wgid] = inc;
+}
+)OpenCLC";
+
+// In this source, each workgroup will generate two values.
+// Every other work item in the workgroup will select either
+// a local or global memory pointer and perform an atomic
+// operation on that.
+const char* KernelSourceVariant = R"OpenCLC(
+kernel void testKernel(global int* globalPtr, local int* localPtr) {
+    int gid = get_global_id(0);
+    int tid = get_local_id(0);
+    int wgid = get_group_id(0);
+    int wgsize = get_local_size(0);
+
+    if (tid == 0) *localPtr = 0;
+
+    barrier(CLK_LOCAL_MEM_FENCE);
+
+    // Initialise the generic pointer to
+    // the global.  Two values are written per WG.
+    generic int* ptr = globalPtr + (wgid * 2);
+
+    // In a workgroup-invariant way, select a localPtr instead.
+    if ((tid % 2) == 0)
+        ptr = localPtr;
+
+    int inc = atomic_fetch_add(ptr, 1);
+
+    barrier(CLK_LOCAL_MEM_FENCE);
+
+    // In the cases where the local memory ptr was used,
+    // save off the final value.
+    if ((tid % 2) == 0)
+        globalPtr[(wgid * 2) + 1] = inc;
+}
+)OpenCLC";
+}
+
+int test_generic_atomics_invariant(cl_device_id deviceID, cl_context context,
+                                   cl_command_queue queue, int num_elements)
+{
+    const auto version = get_device_cl_version(deviceID);
+
+    if (version < Version(2, 0)) return TEST_SKIPPED_ITSELF;
+
+    cl_int err = CL_SUCCESS;
+    clProgramWrapper program = clCreateProgramWithSource(
+        context, 1, &KernelSourceInvariant, nullptr, &err);
+    test_error(err, "clCreateProgramWithSource failed");
+
+    err = clBuildProgram(program, 1, &deviceID, "-cl-std=CL2.0", nullptr,
+                         nullptr);
+    test_error(err, "clBuildProgram failed");
+
+    clKernelWrapper kernel = clCreateKernel(program, "testKernel", &err);
+    test_error(err, "clCreateKernel failed");
+
+    size_t wgSize, retSize;
+    err = clGetKernelWorkGroupInfo(kernel, deviceID,
+                                   CL_KERNEL_PREFERRED_WORK_GROUP_SIZE_MULTIPLE,
+                                   sizeof(wgSize), &wgSize, &retSize);
```

Comment:
This is probably not going to cause an error or the test to fail, but can you please confirm you are intending to query [CL_KERNEL_PREFERRED_WORK_GROUP_SIZE_MULTIPLE](https://registry.khronos.org/OpenCL/specs/3.0-unified/html/OpenCL_API.html#CL_KERNEL_PREFERRED_WORK_GROUP_SIZE_MULTIPLE) here, and not [CL_KERNEL_WORK_GROUP_SIZE](https://registry.khronos.org/OpenCL/specs/3.0-unified/html/OpenCL_API.html#CL_KERNEL_WORK_GROUP_SIZE)?

Reply:
My objective was to select a workgroup size that the device would execute in a single SIMD unit (commonly referred to as "warp" or "wavefront"). I assume that implementers will return that on the PREFERRED workgroup multiple query.
Alternatively, I could query the subgroup size.

Reply:
I see.  FWIW, I'm aware of some devices where this assumption doesn't hold, but I think the test should be fine, regardless.

I'd suggest adding a comment here just to avoid confusion in the future, but not required, up to you.